Additions to markable atomic type

Vtec234 opened the pull request:
The changes we talked about, also accounted for your comments. I note some reasons for changing things in the code comments.

Vtec234 commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -5,33 +5,38 @@ use std::sync::atomic::{self, Ordering};
+/// `mark < std::mem::align_of::<T>()`
 #[derive(Debug)]
-pub struct MarkedAtomic<T> {
+pub struct MarkableAtomic<T> {
```
I changed the name to Markable rather than Marked, since the terminology in concurrency papers is often "marked" for 1, "unmarked" for 0, so the value isn't always "marked".

gereeter replied:
I see your point, but I think `Markable` only makes sense for a boolean tag, where the states are "marked" and "unmarked." In the more general case, I think it makes more sense to think of it as a pointer that always has a mark, just with several possible mark values.

That said, you are right that `Marked` could be confusing. Maybe `Tagged`?

Vtec234 replied:
This seems like a good example of [Wadler's Law](https://wiki.haskell.org/Wadler's_Law) :) Fair, i'll change it to [`TaggedAtomic`](https://en.wikipedia.org/wiki/Tagged_pointer)

Vtec234 commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -48,32 +53,53 @@ fn opt_owned_into_usize<T>(ptr: Option<Owned<T>>, mark: usize) -> usize {
     #[cfg(feature = "nightly")]
-    pub const fn null() -> MarkedAtomic<T> {
-        MarkedAtomic {
+    pub const fn zero() -> MarkableAtomic<T> {
```
I replaced `null` with this, and added a `usize` argument to `null`, since without that it's not possible to initialize a value of e.g. `(0 as *mut _, 20)` safely.

gereeter replied:
This change is good - singling out `0` as a distinguished mark does not seem like a good idea. However, maybe it would be simpler to just leave this function out and make people write `MarkedAtomic::null(0)` - it is only one more character, and is probably clearer.

Vtec234 replied:
That would be better, but notice that for some initializers we need a `const fn`, and a function with a marker argument cannot currently be made `const` due to the `guard_mark` assertion. I don't think assertions in `const fn` will land anytime soon, so we need both functions for now.

Vtec234 commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -5,33 +5,38 @@ use std::sync::atomic::{self, Ordering};
-    let ptr = (val & !(mem::align_of::<T>() - 1)) as *mut T;
-    let mark = val &  (mem::align_of::<T>() - 1);
-    (ptr, mark)
+    let mark = val & (mem::align_of::<T>() - 1);
```
This might or might not be faster, depending on whether `align_of` is evaluated at compile-time.

Vtec234 commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -5,33 +5,38 @@ use std::sync::atomic::{self, Ordering};
-    let mark = val &  (mem::align_of::<T>() - 1);
-    (ptr, mark)
+    let mark = val & (mem::align_of::<T>() - 1);
+    ((val - mark) as *mut T, mark)
```
This might or might not be faster, depending on whether `align_of` is evaluated at compile-time.

gereeter replied:
`align_of` should definitely be evaluated at compile time (if for no other reason that it, as a generic function, gets monomorphised), and I think using only bit operations to extract the parts is clearer anyway.

Vtec234 replied:
Holy crap you're right, on `opt-level=2` `rustc` produces 
```
        push    rbp
        mov     rbp, rsp
        mov     eax, 3
        pop     rbp
        ret
```
for a sum of two `unpack_mark`s. That's pretty cool, i'll revert this change

gereeter commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -48,32 +53,53 @@ fn opt_owned_into_usize<T>(ptr: Option<Owned<T>>, mark: usize) -> usize {
+        }
+    }
+
+    /// Unsafely create a new markable atomic pointer from the given raw pointer.
```
Additions to documentation:

- Mention that this panics if the mark is too large.
- Explain why this is unsafe (like `Atomic`, a `MarkedAtomic` must either be null or be dereferenceable).

gereeter commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -48,32 +53,53 @@ fn opt_owned_into_usize<T>(ptr: Option<Owned<T>>, mark: usize) -> usize {
+    pub fn new(data: T, mark: usize) -> MarkableAtomic<T> {
         guard_mark::<T>(mark);
-        MarkedAtomic {
+        MarkableAtomic {
```
Obviously this code is just what was there previously, but I think a call to `from_raw` would be better here so as to not duplicate code.

That said, it is probably good to call `guard_mark` before allocating a box... so I'm sort of conflicted.

Vtec234 replied:
Eh, it's not a lot of duplication and on a too large mark the panic would leak memory from `Box` if we do `guard_mark` after allocating it, so it's probably fine as it is.

gereeter commented on src/mem/epoch/markable_atomic.rs:
```diff
@@ -237,7 +264,7 @@ impl<T> MarkedAtomic<T> {
     }
 
-    /// Perform a logical "or" on the current mark and the argument `mark` and set the new mark to
+    /// Perform a logical "OR" on the current mark and the argument `mark` and set the new mark to
```
Lowercase "or" matches the standard library documentation. The standard library also doesn't use quotes, though.

Vtec234 commented on src/mem/epoch/tagged_atomic.rs:
```diff
@@ -0,0 +1,295 @@
+    /// # Panics
+    ///
+    /// Panics if `tag >= mem::align_of::<T>()`.
+    pub fn and_tag(&self, tag: usize, ord: Ordering) -> usize {
```
Added `and_tag` because why not, changed the docs to `bitwise`, since it's only a logical operation for booleans and made the two functions return `usize` in case anybody needs it, if not it'll be optimized out anyway.

gereeter replied:
Saying `bitwise` makes sense. I'm fine with returning the old tag, but if so, I'd prefer the functions to be called `fetch_or` and `fetch_and` to match the standard library and to make it very clear which value is fetched. Also, it occurs to me that `xor` could be very useful (for toggling bits in a bitfield, for example), so if it isn't much trouble, could you add that as well?

Vtec234 commented:
I renamed the bitwise functions. Also, there is a bug in `fetch_and_tag` in that it will clear all the higher bits - we have to mask it with `!(mem::align_of::<T>() - 1)` to preserve the pointer. There is a bigger problem though, which I just realized - alignment need not be a power of two, in which all these formulas stop working.

Vtec234 commented:
Okay, got it. What we are really interested in is the number of trailing zeroes in the alignment, which for powers of two is equal to the alignment, otherwise not. I used `trailing_zeros()`, but thanks to the power of constant folding it produces a single constant value.

As for XOR support, it can't be done in a single instruction. We have to mask the pointer bits with the identity values for a particular operation - 0 for OR, 1 for AND. But XOR doesn't have an identity. We could do a CAS loop that reads the tag, XORs it, tags the pointer with the new tag and tries to store it, otherwise repeat, but that's no longer just a `fetch`. If it's not necessary, I wouldn't add it.

**EDIT**: BTW a `guard_tag` in `fetch_and_tag` is technically unnecessary, but we should leave it since too large a tag is still an error regardless of whether it matters for the implementation

gereeter commented:
I thought that alignments were guaranteed to be powers of two - there certainly is [code](https://github.com/rust-lang/rust/blob/f0b42075981d9914c39e848377a3e12f0adf37d7/src/libstd/collections/hash/table.rs#L563) in the standard library that assumes this (though it is still safe if the alignments are not powers of two), and the [custom allocator API](https://github.com/rust-lang/rfcs/blob/a6db7a2e442272501accc86cfd39a58a68a423ab/text/1398-kinds-of-allocators.md#layout-api) seems to guarantee power-of-two alignments. That said, I'm not that fundamentally opposed to being more tolerant, especially since there isn't any runtime cost.

Isn't 0 an identity of xor? The main reason I wanted to support it was that toggling bit fields atomically seems like a real use case.

Yeah, the `guard_tag` calls could all be done away with by masking away any upper bits of the passed-in tag. Their purpose is really more about logical correctness than safety.

One final note (see your comment about Wadler's Law): now that I see the names with `fetch` in them, they look really very confusing. Which value they return is now clear, but the bitwise operation sounds like it is just joining the two verbs `fetch` and `tag`. That is, `fetch_and_tag` sounds like it is simultaneously fetching the old value and setting the tag, which is not true. I'm not sure what the best solution here is. `tag_fetch_and` seems backwards to me, but leaving off the `fetch` also seems wrong.

gereeter commented on src/mem/epoch/tagged_atomic.rs:
```diff
@@ -0,0 +1,301 @@
+/// Returns the power of two that the tag must be strictly
+/// less than. This value is `2.pow(n_unused_bits)`.
+fn tag_ceil<T>() -> usize {
+    0b1 << mem::align_of::<T>().trailing_zeros()
```
Nit: If we keep this support for non-power-of-two alignments, I think it would be slightly cleaner to drop the `0b` prefix. Also, I think the documentation would be slightly clearer as something like:

Returns the largest power of two that divides the alignment of `T`. Any tag placed on a pointer to `T` must be strictly less than this value, so as to not interfere with any bits that the pointer actually uses.

Vtec234 commented:
You're of course right about xor, I got slightly confused by all the ones and zeroes. Added it. Renamed the bitwise fns back to `fetch_fn`. It's short to write and users can just read the short doc to find out which value it acts on.
For `tag_ceil`, we don'y really care about what the alignment divides by (even if it happens to be the right thing), we care about the unused bits, so I wrote that in the doc. This is probably done now.

gereeter commented:
Okay, excellent! It all looks good to me. Thank you again, and sorry if I was being a bit perfectionist.
